src/tests/suite/daos_dist_tx.c

```diff
@@ -0,0 +1,1003 @@
+	MUST(daos_tx_close(th, NULL));
+	MUST(daos_tx_open(arg->coh, &th, 0, NULL));
+
+	dts_buf_render(write_bufs[1], DTX_IO_SMALL);
+	req.arg->expect_result = -DER_EXIST;
+	insert_single_with_flags(dkey, akey, 1, write_bufs[1], DTX_IO_SMALL, th,
+				 &req, DAOS_COND_AKEY_INSERT);
+
+	req.arg->expect_result = 0;
+	insert_single_with_flags(dkey, akey, 1, write_bufs[1], DTX_IO_SMALL, th,
+				 &req, DAOS_COND_AKEY_UPDATE);
+
+	MUST(daos_tx_commit(th, NULL));
+	MUST(daos_tx_close(th, NULL));
+
+	lookup_single_with_rxnr(dkey, akey, 0, fetch_buf, DTX_IO_SMALL,
+				DTX_IO_SMALL * 2, DAOS_TX_NONE, &req);
+	assert_memory_equal(write_bufs[0], fetch_buf, DTX_IO_SMALL);
+	assert_memory_equal(write_bufs[1], &fetch_buf[DTX_IO_SMALL], DTX_IO_SMALL);
+
+	MPI_Barrier(MPI_COMM_WORLD);
+
+	ioreq_fini(&req);
+}
+
+static void
+dtx_10(void **state)
+{
+	test_arg_t	*arg = *state;
+	const char	*dkey = dts_dtx_dkey;
+	const char	*akey = dts_dtx_akey;
+	const char	*dkey2 = "tmp_dkey";
+	const char	*akey2 = "tmp_akey";
+	char		 write_buf[DTX_IO_SMALL];
+	char		 fetch_buf[DTX_IO_SMALL];
+	daos_handle_t	 th = { 0 };
+	daos_obj_id_t	 oid;
+	struct ioreq	 req;
+
+	print_message("DTX with conditional punch\n");
+
+	arg->async = 0;
+	oid = dts_oid_gen(OC_RP_XSF, 0, arg->myrank);
+	ioreq_init(&req, arg->coh, oid, DAOS_IOD_SINGLE, arg);
+
+	dts_buf_render(write_buf, DTX_IO_SMALL);
+	insert_single(dkey, akey, 0, write_buf, DTX_IO_SMALL, DAOS_TX_NONE,
+		      &req);
+
+	MPI_Barrier(MPI_COMM_WORLD);
+
+	MUST(daos_tx_open(arg->coh, &th, 0, NULL));
+
+	req.arg->expect_result = -DER_NONEXIST;
+	punch_akey_with_flags(dkey2, akey2, th, &req, DAOS_COND_PUNCH);
+
+	req.arg->expect_result = 0;
+	punch_akey_with_flags(dkey, akey, th, &req, DAOS_COND_PUNCH);
+
+	req.arg->expect_result = -DER_NONEXIST;
+	punch_dkey_with_flags(dkey2, th, &req, DAOS_COND_PUNCH);
+
+	req.arg->expect_result = 0;
+	punch_dkey_with_flags(dkey, th, &req, DAOS_COND_PUNCH);
+
+	MUST(daos_tx_commit(th, NULL));
+
+	lookup_single(dkey, akey, 0, fetch_buf, DTX_IO_SMALL,
+		      DAOS_TX_NONE, &req);
+	assert_int_equal(req.iod[0].iod_size, 0);
+
+	ioreq_fini(&req);
+	MUST(daos_tx_close(th, NULL));
+
+	MPI_Barrier(MPI_COMM_WORLD);
+}
+
+static void
+dtx_11(void **state)
+{
+	test_arg_t	*arg = *state;
+	const char	*dkey = dts_dtx_dkey;
+	const char	*akey = dts_dtx_akey;
+	const char	*dkey2 = "tmp_dkey";
+	const char	*akey2 = "tmp_akey";
+	char		 write_buf[DTX_IO_SMALL];
+	char		 fetch_buf[DTX_IO_SMALL];
+	daos_handle_t	 th = { 0 };
+	daos_obj_id_t	 oid;
+	struct ioreq	 req;
+
+	print_message("Read only transaction\n");
+
+	arg->async = 0;
+	oid = dts_oid_gen(OC_RP_XSF, 0, arg->myrank);
+	ioreq_init(&req, arg->coh, oid, DAOS_IOD_SINGLE, arg);
+
+	dts_buf_render(write_buf, DTX_IO_SMALL);
+	insert_single(dkey, akey, 0, write_buf, DTX_IO_SMALL, DAOS_TX_NONE,
+		      &req);
+
+	insert_single(dkey2, akey, 0, write_buf, DTX_IO_SMALL, DAOS_TX_NONE,
+		      &req);
+
+	MUST(daos_tx_open(arg->coh, &th, DAOS_TF_RDONLY, NULL));
```

Comment:
I am a little worried about allowing this.
do we guarantee that a readonly transaction is going to be available throughout its lifetime to be readable?

Reply:
ok we discussed this on the call.

Reply:
As Liwei's idea , we have some solution for that although I am not sure whether it is efficient enough or not.

Reply:
As Liwei's idea , we have some solution for that although I am not sure whether it is efficient enough or not.

Reply:
I mean the solution for the race between transactional read and vos aggregation.